src/eos-updater-apply.c

```diff
@@ -73,41 +77,592 @@ get_test_osname (void)
+
+      flatpak_ref = flatpak_ref_parse (ref, error);
+
+      if (!flatpak_ref)
+        return NULL;
+
+      g_ptr_array_add (flatpaks,
+                       g_object_new (FLATPAK_TYPE_REMOTE_REF,
+                                     "remote-name", remote,
+                                     "name", flatpak_ref_get_name (flatpak_ref),
+                                     "kind", flatpak_ref_get_kind (flatpak_ref),
+                                     NULL));
+    }
+
+  return g_steal_pointer (&flatpaks);
+}
+
+typedef gpointer (*PtrCopyFunc) (gpointer src);
+
+static GPtrArray *
+concat_ptrarrays (GPtrArray *left,
+                  GPtrArray *right,
+                  PtrCopyFunc copy_func,
+                  GDestroyNotify destroy_func)
+{
+  gsize left_size = left ? left->len : 0;
+  gsize right_size = right ? right->len: 0;
+  GPtrArray *ptr_array = g_ptr_array_new_full (left_size + right_size, destroy_func);
+  gsize i;
+
+  for (i = 0; i < left_size; ++i)
+    g_ptr_array_add (ptr_array, copy_func (g_ptr_array_index (left, i)));
+
+  for (i = 0; i < right_size; ++i)
+    g_ptr_array_add (ptr_array, copy_func (g_ptr_array_index (right, i)));
+
+  return ptr_array;
+}
+
+static GPtrArray *
+list_flatpak_refs_from_d_directory (GFile        *directory,
+                                    GCancellable *cancellable,
+                                    GError       **error)
+{
+  g_autofree gchar *autoinstall_flatpak_contents = NULL;
+  g_autofree gchar *autoinstall_flatpak_file_path = NULL;
+  g_auto(GStrv) autoinstall_flatpak_lines = NULL;
+  g_autoptr(GFileEnumerator) autoinstall_d_enumerator = NULL;
+  g_autoptr(GPtrArray) autoinstall_flatpaks = NULL;
+
+  /* Repository checked out, read all files in orer and build up a list
+   * of flatpaks to auto-install */
+  autoinstall_d_enumerator = g_file_enumerate_children (directory,
+                                                        "standard::*",
+                                                        G_FILE_QUERY_INFO_NONE,
+                                                        cancellable,
+                                                        error);
+
+  if (!autoinstall_d_enumerator)
+    return NULL;
+
+  while (TRUE)
+    {
+      g_autoptr(GPtrArray) flatpak_refs = NULL;
+      g_autoptr(GPtrArray) last_flatpak_refs = NULL;
+      GFile *file;
+
+      if (!g_file_enumerator_iterate (autoinstall_d_enumerator,
+                                      NULL, &file,
+                                      cancellable, error))
+        return NULL;
+
+      if (!file)
+        break;
+
+      flatpak_refs = read_flatpak_refs_from_file (file, cancellable, error);
+
+      if (!flatpak_refs)
+        return NULL;
+
+      /* We need to take a binding here to flatpak_refs by stealing the pointer
+       * so that it will be freed later */
+      last_flatpak_refs = g_steal_pointer (&autoinstall_flatpaks);
+      autoinstall_flatpaks = concat_ptrarrays (last_flatpak_refs,
```

Comment:
Yeah, I had that though too, though I do prefer the functional style of `concat_ptrarrays`. Since performance is a potential concern here, I'll make `read_flatpak_refs_from_file` extend an existing GPtrArray (which I suppose at least reduces the cost to nlogn so long as GPtrArray does allocation the way I think it does).

Reply:
(side note: I wonder if GXLib has lazy-pointer-arrays, that'd help to alleviate the cost here ;-) )